Rolling tombstones clean up

ArthurSens opened the pull request:
Today while cleaning tombstones, new blocks are created and the old blocks remain in disk until the end of the compaction operation. Only after that, when all the blocks are being reloaded, is when all the old blocks are deleted.

This PR changes this behavior, reloading the blocks after each block compaction. The goal is to remove deletable blocks after every compaction and, thus, decrease disk space used when cleaning tombstones

Fixes #7957 

<!--
    Don't forget!
    
    - If the PR adds or changes a behaviour or fixes a bug of an exported API it would need a unit/e2e test.
    
    - Where possible use only exported APIs for tests to simplify the review and make it as close as possible to an actual library usage.
    
    - No tests are needed for internal implementation changes.
    
    - Performance improvements would need a benchmark test to prove it.
    
    - All exposed objects should have a comment.
    
    - All comments should start with a capital letter and end with a full stop.
 -->

brian-brazil commented:
You've got a test failure on Windows.

codesome commented on tsdb/db.go:
```diff
@@ -1499,9 +1499,13 @@ func (db *DB) CleanTombstones() (err error) {
 			return err
 		} else if uid != nil { // New block was created.
 			newUIDs = append(newUIDs, *uid)
+			if err = errors.Wrap(db.reload(), "reload blocks"); err != nil {
+				return err
+			}
```
To be consistent with rest of the code
```suggestion
			if err = db.reload(); err != nil {
				return errors.Wrap(err, "reload blocks")
			}
```

codesome reviewed (commented):
The tests are prolly failing because after calling the reload() inside the loop it is now working with old set of blocks

ArthurSens commented:
I've made some changes, trying to assure I'm not looping over old blocks but looks like it was not enough :disappointed: 
Any tips are welcome

brian-brazil commented on tsdb/db.go:
```diff
@@ -1476,32 +1476,42 @@ func (db *DB) CleanTombstones() (err error) {
+					return errors.Wrapf(er, "clean tombstones: %s", b.Dir())
+				} else if uid != nil { // New block was created.
+					if err = db.reload(); err != nil {
+						// db.reload will try to delete old block
```
It's probably best to just fail here, rather than messing around further after an error has occured.

ArthurSens replied:
Please correct me if I'm wrong, but b.CleanTombstones will create a new block, reload will try to delete the old one. If it fails to delete and we leave the two blocks there, we will have overlapping timeseries.

On this part of the code, I'm basically refactoring the deferred function that deleted all new blocks after a failed reload. Since we are deleting one by one now, we can do this validation one by one as well

ArthurSens replied:
After another read on the reload() function, I've noticed that parent(old) blocks are [marked as deletable](https://github.com/prometheus/prometheus/blob/d6fa7f72a627564966e9429d3c106685fbb7c340/tsdb/db.go#L922-L935) already.

I was wondering if, after failing to delete an old block, it is really a problem to not delete the new compacted ones (today the new ones are being deleted after failing to delete the old ones). Won't the reload() function be called multiple times on other operations and, thus, there will be several other retries to delete the old blocks?

brian-brazil replied:
Exactly, there's no need to do much beyond returning an error. If this fails we're already in a bad state.

ArthurSens replied:
After a really dirty debugging and a lot of "fmt.Println" 🥵 I've finally found out that tests were failing after I removed this error handling (at Linux at least). A new compacted block is created, it then fails to delete the old one due to overlapping block sequence. Currently, tests expect that the new compacted block is deleted after a failed reload. 

--

So, just to make things a little bit clearer for me... Should I really leave new blocks there after failing to remove old ones, then change tests accordingly? Or should I keep this behavior?

Personally, I would prefer to keep this behavior, I don't see why leaving both the new compacted block and its parent there... I think one of the two should be deleted.

WDYT @brian-brazil @codesome @bwplotka?

brian-brazil commented on tsdb/db.go:
```diff
@@ -1476,32 +1476,42 @@ func (db *DB) CleanTombstones() (err error) {
-		} else if uid != nil { // New block was created.
-			newUIDs = append(newUIDs, *uid)
+		blocks = db.Blocks()
+		for _, b := range blocks {
```
blocks will be invalidated after the reload, so you should check that the b is still in the current list.

ArthurSens replied:
Sorry, I'm not sure I understood your comment. The `break` at the end of the inner loop will make sure that I'm not looping over the same blocks after a reload.

ArthurSens commented on tsdb/db.go:
```diff
@@ -1476,32 +1476,42 @@ func (db *DB) CleanTombstones() (err error) {
+
+					// This is necessary because the blocks change after a successful db.reload()
+					// We need to loop over the new set of blocks
+					continue
```
Yep I did :facepalm:

codesome commented on tsdb/block.go:
```diff
@@ -568,6 +568,9 @@ func (pb *Block) CleanTombstones(dest string, c Compactor) (*ulid.ULID, error) {
 		return nil, err
 	}
+
+	pb.Close()
```
We should not be closing the block here, we should let the DB reload do the cleanup.

ArthurSens replied:
That was just an experimental thing on the last commit... I will just revert the last commit, but the tests were not passing before as well.

codesome reviewed (commented):
Other changes actually look fine with 1 small issue. Let fix that and see if it works.

bwplotka commented:
Can we fix tests? (:

ArthurSens commented on tsdb/db.go:
```diff
@@ -1510,35 +1510,45 @@ func (db *DB) CleanTombstones() (err error) {
+						// Delete directory of new block
+						dir := filepath.Join(db.Dir(), uid.String())
+						if err := os.RemoveAll(dir); err != nil {
+							fmt.Println("Failed to remove dir: ", dir, " err: ", err)
```
Windows tests tell that the new block couldn't be deleted because it is opened somewhere. Couldn't figure out how to close it though

ArthurSens commented:
I took a different approach in these last commits. Resolved all reviews since they didn't seem to apply anymore. 
Ready for another round of reviews :) 

And about the fuzzit_regression test failure, I'm not sure why it failed. Should I be doing something about it?

codesome commented on tsdb/db.go:
```diff
@@ -1505,34 +1505,38 @@ func (db *DB) CleanTombstones() (err error) {
-			}
-		}
-	}()
+	// We can momentarily allow overlapping blocks in this particular case.
+	// This will allow the reload function to replace old blocks with the new ones.
+	// Notice that the the final result will be overlapping blocks, but with no tombstones.
+	if !db.opts.AllowOverlappingBlocks {
+		db.opts.AllowOverlappingBlocks = true
+		defer func() { db.opts.AllowOverlappingBlocks = false }()
+	}
```
We don't need this. And the final result is not overlapping blocks because the `reload()` will remove the old block.

codesome commented on tsdb/db.go:
```diff
@@ -1505,34 +1505,38 @@ func (db *DB) CleanTombstones() (err error) {
+				// We need to loop over the new set of blocks.
+				break
+			}
+			return errors.Wrap(err, "reload blocks")
```
We should be deleting the new block in case the reload fails.

ArthurSens replied:
What I'm failing to understand is why adding this logic here fails the windows tests :disappointed:

ArthurSens replied:
The new test shows that this problem doesn't occur on partial failures (some blocks get replaced and some dont) at least

codesome commented on tsdb/db.go:
```diff
@@ -1505,34 +1505,38 @@ func (db *DB) CleanTombstones() (err error) {
 
-	if err := db.reloadBlocks(); err != nil {
-		return errors.Wrap(err, "reload blocks")
+			pb.meta.Compaction.Deletable = true
```
We don't need this since the new block has the old block as a parent and hence will be deleted.

ArthurSens replied:
That's what I thought too, but to my surprise, they are not deleted without this line

codesome replied:
I saw your comment in the email saying `That's what I thought too, but to my surprise, they are not deleted without this line`, and I think that is possible! If the tombstones covered the entire block, then there would be no resultant block to tell the parent. So having this here makes sense.

codesome replied:
You can add a comment describing why it is required (the reason that I stated above)

ArthurSens replied:
Yeah, I've written that but then remembered that on test "TestTombstoneClean" it works just fine without it. I deleted the comment just because I wanted to investigate a little bit more. 

What you mention above is exactly what happens on "TestTombstoneCleanFail", the new block is created without a parent, and the old block is empty :+1:

ArthurSens commented:
Follow up from prometheus office hours

Check for meta numSamples == 0 on pb.CleanTombstones and not return uid

roidelapluie commented on tsdb/db_test.go:
```diff
@@ -1110,10 +1110,11 @@ func TestTombstoneCleanFail(t *testing.T) {
 	// totalBlocks should be >=2 so we have enough blocks to trigger compaction failure.
 	totalBlocks := 2
 	for i := 0; i < totalBlocks; i++ {
+		// Overlapping blocks will fail right at the start
```
Comment must end with a `.`

roidelapluie commented on tsdb/db_test.go:
```diff
@@ -1110,10 +1110,11 @@ func TestTombstoneCleanFail(t *testing.T) {
 	var expectedBlockDirs []string
 
-	// Create some empty blocks pending for compaction.
+	// Create some empty blocks with overlapping mint and maxt
```
Comment must end with a `.`

roidelapluie commented on tsdb/db_test.go:
```diff
@@ -1126,6 +1127,42 @@ func TestTombstoneCleanFail(t *testing.T) {
+
+// TestTombstoneCleanPartialFail tests that a failing TombstoneClean doesn't leave any blocks behind.
+// It differs from TestTombstoneCleanFail because some of the blocks are succesfully cleaned.
+// The blocks that got cleaned should be replaced, while the failed one shouldn't
```
Comment must end with a `.`

ArthurSens commented:
@roidelapluie @bwplotka, looks like that our idea during Prometheus office hours still wasn't enough for windows :(

ArthurSens commented on tsdb/block.go:
```diff
@@ -554,15 +555,22 @@ func (pb *Block) CleanTombstones(dest string, c Compactor) (*ulid.ULID, error) {
+		return nil, false, err
+	}
+
+	// Cleaning a block with no timeseries shouldn't return a new block
+	// Since it's empty, it will already be deleted on next truncation
```
Comments should end with `.`

roidelapluie commented:
> @roidelapluie @bwplotka, looks like that our idea during Prometheus office hours still wasn't enough for windows :(

You might want to delete the block straight away also

roidelapluie commented on tsdb/db.go:
```diff
@@ -1505,34 +1505,41 @@ func (db *DB) CleanTombstones() (err error) {
+			// then there would be no resultant block to tell the parent.
+			pb.meta.Compaction.Deletable = safeToDelete
+			cleanUpCompleted = false
+			if err = db.reload(); err == nil { // Will try to delete old block.
```
I think this is not enough to delete the block?

ArthurSens replied:
Reloading is supposed to delete obsolete blocks, such as the ones that were already cleaned, and add the new block to the db object.

If we fail to delete an obsolete block, for whatever reason, then we should delete the new one. This is not done on reload(), so there is extra logic after that to handle this situation.

ArthurSens replied:
Some fmt.print was added to see if the error expected by the test was indeed the one we,re receiving, which it is indeed.

So now I think that something during reload() is holding the new block open. I'm trying to debug this but it's quite complicated since it behaves differently on Linux 😅

ArthurSens commented:
Okey! It should be good for another round of review now 🥵

* Previously `TestTombstoneClean` was [creating overlapping blocks](https://github.com/prometheus/prometheus/blob/e5478983c43cb839a77c144f953b9ea25783b995/tsdb/db_test.go#L1117) (which is impossible on normal operations). That would trigger some [issues at reload()](https://github.com/prometheus/prometheus/blob/e5478983c43cb839a77c144f953b9ea25783b995/tsdb/db.go#L1025-L1029) if overlapping blocks were not allowed, but a [MockCompactorFailing](https://github.com/prometheus/prometheus/blob/e5478983c43cb839a77c144f953b9ea25783b995/tsdb/db_test.go#L1131) triggers a different failure before we get to that.
* New `TestTombstoneClean` does not create overlapping blocks, since we call reload() early and multiple times. It checks for partial replacement, where some blocks are replaced and some aren't.
* A new test case was added to make sure a block resultant from CleanTombstones() is deleted if the tombstones covered the whole block.



Obs: Not sure why fuzzit_regression is failing. Is it flakiness? :confused:

roidelapluie commented:
For fuzzier you can revise.

ArthurSens commented:
> For fuzzier you can revise.

Hey, sorry for the late response! What do you mean by revise? Translator didn't help me that much :point_right::point_left: 

I'm reading more about fuzzit regression tests, but I'm not sure if I understood their purpose. I may be wrong, but the test output doesn't look related to changes on this PR :thinking:

roidelapluie commented:
oooh.. rebase, not revise, sorry

roidelapluie commented:
> I'm reading more about fuzzit regression tests, but I'm not sure if I understood their purpose.

https://en.wikipedia.org/wiki/Fuzzing

Shortly, we feed a lot of input data that might be passed by users, large, small, regexes, errors, ... and see if Prometheus crashes or bugs in any way.

codesome commented on tsdb/block.go:
```diff
@@ -554,15 +556,25 @@ func (pb *Block) CleanTombstones(dest string, c Compactor) (*ulid.ULID, error) {
+	// Cleaning tombstones shouldn't return anything if the result is an empty block.
+	// Just indicating that it's safe to delete the old block is enough.
+	if meta.Stats.NumSeries == 0 {
+		dir := filepath.Join(dest, uid.String())
+		if err := os.RemoveAll(dir); err != nil {
+			level.Error(pb.logger).Log("msg", "failed to delete empty block during `CleanTombstones`", "dir", dir, "err", err)
+		}
```
Empty block will never be written, so this part is not required.

codesome commented on tsdb/db.go:
```diff
@@ -1522,34 +1522,40 @@ func (db *DB) CleanTombstones() (err error) {
+			// then there would be no resultant block to tell the parent.
+			pb.meta.Compaction.Deletable = safeToDelete
+			cleanUpCompleted = false
+			if err = db.reload(); err == nil { // Will try to delete old block.
```
```suggestion
			if err = db.reloadBlocks(); err == nil { // Will try to delete old block.
```

codesome commented on tsdb/db.go:
```diff
@@ -1522,34 +1522,40 @@ func (db *DB) CleanTombstones() (err error) {
+				break
+			}
+
+			// Delete new block if it was created.
+			if uid != nil {
```
Feels like there could be a race here since we only know the reload failed. It is also possible the time/size retention kicked in and deletion of those block failed and the one we wanted to delete was deleted. So we should reliably verify that the parent still exists else we lose data. Maybe call `db.Blocks()` here and see if the parent still exists and then delete.

Can you also add a test case for this?

ArthurSens replied:
If time/size retention kicks in and deletes the parent in the middle of cleanTombstones(), I don't see a problem in cleaning the new block as well since they also should be cleaned the retention limits, don't they?

brian-brazil replied:
Shouldn't we already be holding a lock so only one compaction-y event can be ongoing at once?

ArthurSens replied:
There is a lock over here already:
https://github.com/prometheus/prometheus/blob/e483cff61ff1b68e8925db1059393c4d36af9df5/tsdb/db.go#L1034-L1037

But yeah, I think it is possible to have a race here where cleanTombstones and a deletion cause by retention limits can happen at the same time:
https://github.com/prometheus/prometheus/blob/e483cff61ff1b68e8925db1059393c4d36af9df5/tsdb/db.go#L1053-L1055
There is no problem on Linux with this, but I think that windows would complain if we try to delete a file that does not exist anymore.

Would it be safe to move the unlock from line 1037 to line 1056? We would lock this whole deletion process.

brian-brazil replied:
That's a bug then, we shouldn't be doing operations like that in parallel.

ArthurSens replied:
Does the last commit solve the race?

codesome replied:
`uid != nil` does not signify a non-empty block, it should be `uid != nil && *uid != ulid.ULID{}` since thats what compactor returns.

codesome commented on tsdb/db_test.go:
```diff
@@ -1030,7 +1030,7 @@ func TestTombstoneClean(t *testing.T) {
 		// Delete the ranges.
 
-		// create snapshot
+		// create snapshot.
```
While you are on it, we can capitalize the first letter :) (also the other comment)

ArthurSens commented:
Just a side note...

The CI tests that failed, passed locally. I'll investigate why it failed on CI, I just couldn't find time to do so yet :(

roidelapluie commented:
To reproduce locally make sure you use `go test -race`.

beorn7 commented:
@codesome could you have a final look at this? It seems ready for review.

codesome commented on tsdb/db.go:
```diff
@@ -966,6 +966,12 @@ func (db *DB) reloadBlocks() (err error) {
 	}()
 
+	// Now that we reload TSDB every minute, there is high chance for race condition with a reload
+	// triggered by cleanTombstones(). We need to lock the reload to avoid the situation where
```
```suggestion
	// triggered by CleanTombstones(). We need to lock the reload to avoid the situation where
```

codesome commented on tsdb/db.go:
```diff
@@ -966,6 +966,12 @@ func (db *DB) reloadBlocks() (err error) {
 
+	// Now that we reload TSDB every minute, there is high chance for race condition with a reload
+	// triggered by cleanTombstones(). We need to lock the reload to avoid the situation where
+	// a normal reload and cleanTombstones try to delete the same block.
```
```suggestion
	// a normal reload and CleanTombstones try to delete the same block.
```

codesome commented on tsdb/db_test.go:
```diff
@@ -1137,10 +1185,76 @@ func TestTombstoneCleanFail(t *testing.T) {
+	require.Equal(t, len(intersection(oldBlockDirs, actualBlockDirs)), len(actualBlockDirs)-1)
+}
+
+// TestTombstoneCleanRetentionLimitsRace tests that a cleanTombstones operation
```
```suggestion
// TestTombstoneCleanRetentionLimitsRace tests that a CleanTombstones operation
```

codesome commented on tsdb/db_test.go:
```diff
@@ -1137,10 +1185,76 @@ func TestTombstoneCleanFail(t *testing.T) {
+	opts := DefaultOptions()
+	var wg sync.WaitGroup
+
+	// We want to make sure that a race doesn't happen when a normal reload and a cleanTombstones()
```
```suggestion
	// We want to make sure that a race doesn't happen when a normal reload and a CleanTombstones()
```

codesome reviewed (commented):
LGTM with some nits and checking a non empty block :)

ArthurSens commented:
Thanks for having another look!

I've addressed your comments, but your suggestion to use `uid != nil && *uid != ulid.ULID{}` does not compile :P, I've added parenthesis but it's unclear to me if it has the expected behavior.

Also, windows test is failing with:
```
bash: go: command not found
bash: line 1: go: command not found
```
Looks like flakiness 🤔

ArthurSens commented:
Tests are passing again after a rebase

codesome commented:
> I've addressed your comments, but your suggestion to use uid != nil && *uid != ulid.ULID{} does not compile :P, I've added parenthesis but it's unclear to me if it has the expected behavior.

Not sure why it did not compile, I was able to run this test locally

```go
func TestULID(t *testing.T) {
	uid := &ulid.ULID{}
	require.False(t, uid != nil && *uid != ulid.ULID{})
	require.False(t, uid != nil && *uid != (ulid.ULID{}))

	(*uid)[10] = 25
	require.True(t, uid != nil && *uid != ulid.ULID{})
	require.True(t, uid != nil && *uid != (ulid.ULID{}))
}
```

codesome reviewed (approved):
Thanks!

ArthurSens commented:
> Not sure why it did not compile, I was able to run this test locally
> 
> ```go
> func TestULID(t *testing.T) {
> 	uid := &ulid.ULID{}
> 	require.False(t, uid != nil && *uid != ulid.ULID{})
> 	require.False(t, uid != nil && *uid != (ulid.ULID{}))
> 
> 	(*uid)[10] = 25
> 	require.True(t, uid != nil && *uid != ulid.ULID{})
> 	require.True(t, uid != nil && *uid != (ulid.ULID{}))
> }
> ```

That is strange indeed 🤔 
![image](https://user-images.githubusercontent.com/24193764/108208931-6bf7a180-7108-11eb-818b-62397ea09160.png)

![image](https://user-images.githubusercontent.com/24193764/108209111-a95c2f00-7108-11eb-81dd-606a4cc8a748.png)

ArthurSens commented:
Anyway, thanks for your patience and guidance during these 5 months @codesome @bwplotka @roidelapluie 🙂
